tests/unit/cli/test_helpers.py

```
@@ -919,3 +942,262 @@ async def test_returned_list_includes_non_url_sources_like_research_reports(self
+        # Must retry — the URL being pre-existing isn't proof our import wrote.
+        assert client.research.import_sources.await_count == 2
+        mock_sleep.assert_awaited_once_with(5)
+        assert imported == [{"id": "src_existing", "title": "Old"}]
+
+    @pytest.mark.asyncio
+    async def test_report_only_import_bounded_retries_on_persistent_timeout(self):
+        """Report-only deep-research imports (no URLs) can't use the URL-match
+        verification path. To bound the worst-case duplicate inflation, the
+        retry loop must give up after a small number of attempts rather than
+        burning the full ``max_elapsed`` budget — otherwise a persistent
+        timeout still produces 5–6× duplicate reports.
```

Ruff (0.15.12)

[warning] Docstring contains ambiguous `–` (EN DASH). Did you mean `-` (HYPHEN-MINUS)?

(RUF002)